### ISSUE TITLE
`VecMap::as_slice` causes segmentation fault

### ISSUE THREAD

**Issue by extremeandy** [martinohmann/vecmap-rs#18]
I am seeing a `segmentation fault` error calling `VecMap::as_slice()`.

I think the memory layout of `struct Slot<K, V>` will not be the same as `(K, V)` unless you use `#repr(packed)`.

```rust
pub fn as_slice(&self) -> &[(K, V)] {
    // SAFETY: `&[Slot<K, V>]` and `&[(K, V)]` have the same memory layout.
    unsafe { &*(self.base.as_slice() as *const [Slot<K, V>] as *const [(K, V)]) }
}
```

I am happy to make a pull request to fix this and add some tests if you like?

**Comment by khigia** [comment]
A `repr(packed)` may not solve the issue. The tuple layout does not seems to be packed from this doc (https://doc.rust-lang.org/reference/type-layout.html#tuple-layout). Maybe an easier way is to replace `Slot` by something easier to transmute to pair.
```
struct Slot<K, V>{key:K,val:V}

#[repr(packed)]
struct SlotP<K, V>{key:K,val:V}

struct SlotT<K, V>(K,V);
struct SlotW<K, V>((K,V));


static_assertions::assert_eq_size!(Slot<i64, String>, (i64, String));
static_assertions::assert_eq_size!(SlotP<i64, String>, (i64, String));
static_assertions::assert_eq_size!(SlotT<i64, String>, (i64, String));
static_assertions::assert_eq_size!(SlotW<i64, String>, (i64, String));

static_assertions::assert_eq_size!(Slot<i64, u8>, (i64, u8));
// static_assertions::assert_eq_size!(SlotP<i64, u8>, (i64, u8));  <-- fail
static_assertions::assert_eq_size!(SlotT<i64, u8>, (i64, u8));
static_assertions::assert_eq_size!(SlotW<i64, u8>, (i64, u8));

static_assertions::assert_eq_align!(Slot<i64, u8>, (i64, u8));
// static_assertions::assert_eq_align!(SlotP<i64, u8>, (i64, u8));  <-- fail
static_assertions::assert_eq_align!(SlotT<i64, u8>, (i64, u8));
static_assertions::assert_eq_align!(SlotW<i64, u8>, (i64, u8));

```

**Comment by martinohmann** [comment]
Yikes, this is bad. Can you provide a reproducer for this segfault? I suspect this might be caused by field reordering, e.g. when tuples like `(u8, i64)` are involved.

Changing `Slot` to `SlotT` will probably not help as this is also subject to field reordering AFAIK, but something like `SlotW` might fix the issue.

In any case this calls for better test coverage for `as_slice` with various tuple combinations. Maybe miri could have detected this, but I never used it until now. Maybe I should...

If you want to work on a fix please go ahead. I will probably not be able to work on this until the new year :).

**Comment by martinohmann** [comment]
The safest bet should be to just use a transparent wrapper around a tuple to ensure the layout is always the same as `(K, V)`:

```rust
#[repr(transparent)]
struct Slot<K, V>((K, V));
```

**Comment by martinohmann** [comment]
@extremeandy thank you for reporting this! I just released `vecmap-rs` 0.1.14 which fixes the unsoundness. Merry christmas 😉.
